sklearn/inspection/_h_statistic.py

```diff
@@ -0,0 +1,297 @@
+    >>> H
+    {'feature_pairs': [(3, 8), (3, 2), (8, 2)],
+     'h_squared_pairwise': array([0.00985985, 0.00927104, 0.03439926]),
+     'numerator_pairwise': array([ 1.2955532 ,  1.2419687 , 11.13358385]),
+     'denominator_pairwise': array([131.39690331, 133.96210997, 323.6576595 ])}
+
+    >>> # Interpretation: For features (8, 2), 3.4% of the joint effect variability
+    >>> # comes from their interaction. These two features also have strongest absolute
+    >>> # interaction across pairs ("numerator_pairwise").
+    """
+    check_is_fitted(estimator)
+
+    if not (is_classifier(estimator) or is_regressor(estimator)):
+        raise ValueError("'estimator' must be a fitted regressor or classifier.")
+
+    if is_classifier(estimator) and isinstance(estimator.classes_[0], np.ndarray):
+        raise ValueError("Multiclass-multioutput estimators are not supported")
+
+    # Use check_array only on lists and other non-array-likes / sparse. Do not
+    # convert DataFrame into a NumPy array.
+    if not (hasattr(X, "__array__") or sparse.issparse(X)):
+        X = check_array(X, force_all_finite="allow-nan", dtype=object)
+
+    if sample_weight is not None:
+        sample_weight = _check_sample_weight(sample_weight, X)
+
+    # Usually, the data is too large and we need subsampling
+    if X.shape[0] > n_max:
+        row_indices = sample_without_replacement(
+            n_population=X.shape[0], n_samples=n_max, random_state=random_state
+        )
+        X = _safe_indexing(X, row_indices, axis=0)
+        if sample_weight is not None:
+            sample_weight = _safe_indexing(sample_weight, row_indices, axis=0)
+    else:
+        X = X.copy()
+
+    if features is None:
+        features = feature_indices = np.arange(X.shape[1])
+    else:
+        feature_indices = np.asarray(
+            _get_column_indices(X, features), dtype=np.intp, order="C"
+        ).ravel()
+
+    # CALCULATIONS
+    pd_univariate = []
+    for idx in feature_indices:
+        pd_univariate.append(
+            _calculate_pd_over_data(
+                estimator, X=X, feature_indices=[idx], sample_weight=sample_weight
+            )
+        )
+
+    num = []
+    denom = []
+
+    for j, k in itertools.combinations(range(len(feature_indices)), 2):
+        pd_bivariate = _calculate_pd_over_data(
+            estimator,
+            X=X,
+            feature_indices=feature_indices[[j, k]],
+            sample_weight=sample_weight,
+        )
+        num.append(
+            np.average(
+                (pd_bivariate - pd_univariate[j] - pd_univariate[k]) ** 2,
+                axis=0,
+                weights=sample_weight,
+            )
+        )
+        denom.append(np.average(pd_bivariate**2, axis=0, weights=sample_weight))
+
+    num = np.array(num)
+    num[np.abs(num) < eps] = 0  # Round small numerators to 0
+    denom = np.array(denom)
+    h2_stat = np.divide(num, denom, out=np.zeros_like(num), where=denom > 0)
```

Comment:
> Should this emit a warning in case some value of denom is zero?

It turns inf/nan results from division by 0 to 0. When both features do not have any effect, then both the numerator and denominator are zero. In this case, it is convenient to get a 0 (= "there is no interaction"). We can change this of course.

```
num = np.array([1, 0, 1])
denom = np.array([1, 0, 0])

np.divide(num, denom, out=np.zeros(shape=3))
# Gives array([ 1., nan, inf])

np.divide(num, denom, out=np.zeros(shape=3), where=denom > 0)
# Gives array([1., 0., 0.])
```